api/queue/work/model_service_deployment.go

```diff
@@ -111,8 +111,15 @@ func (depl *ModelServiceDeployment) Deploy(job *queue.Job) error {
 
 		// record the deployment result
 		deployment.UpdatedAt = time.Now()
-		if _, err := depl.DeploymentStorage.Save(deployment); err != nil {
-			log.Warnf("unable to update deployment history", err)
+		if deployment.IsSuccess() {
+			if err := depl.DeploymentStorage.OnDeploymentSuccess(deployment); err != nil {
```

Comment:
im at dilemma for these, given the current state of how it is written it seems fine. So just leaving a comment and not expecting any changes for now.

I would expect the service to use the gorm.DB directly to start transaction but its abstracted away for now or the storage to expose transactional methods.

Personally I think there are many side effects and unclear segregation of responsibilities. I would expect storage layer to own "How to save to db" and the service to own "How and what to save to db". 
Its within the storage layer logic that all old deployment must be updated. It can expose a method to return all past deployment, the service layer updates the deployment and logic, then pass a list of deployment for the storage. The storage would simply persist the list.

Imagine using another storage, it would assume the storage should decide the old deployment gets updated or not.

Now it looks like `Save` is for failure and `OnDeploymentSuccess` for success case. It is also not obvious from these.

Reply:
Thanks, Leon for sharing your concern and idea. I'm agree that the separation of concern between service and storage layers in Merlin codebase is not clear enough leading the confusion on how to implement both of them. I also agree that service is responsible for "what to save" and storage layer is responsible for "how to save the data".

However, this might need more time and thinking to achieve the best solution, so for now I'll create GitHub issue to track this: https://github.com/caraml-dev/merlin/issues/490 so we can revisit it later.

Reply:
yep not expecting to be done here at all